classes/file_storage.php

```diff
@@ -546,19 +586,46 @@ private static function exportFileTree($target, $contextid, $filearea, $filepath
         $fs = get_file_storage();
         $files = $fs->get_directory_files($contextid, 'mod_hvp', $filearea, $itemid, $filepath, true);
 
-        foreach ($files as $file) {
-            // Correct target path for file.
-            $path = $target . str_replace($filepath, '/', $file->get_filepath());
-
-            if ($file->is_directory()) {
-                // Create directory.
-                $path = rtrim($path, '/');
-                if (!file_exists($path)) {
-                    mkdir($path, 0777, true);
+        if ($filearea === 'libraries' && $fs->file_exists(
+            $contextid,
+            'mod_hvp',
+            $filearea,
+            $itemid,
+            $filepath,
+            'lib-export.zip'
+        )) {
+            // Libraries may have a precompiled zip to extract(
+            $file = $fs->get_file(
+                $contextid,
+                'mod_hvp',
+                $filearea,
+                $itemid,
+                $filepath,
+                'lib-export.zip'
+            );
+            $temppath = $file->copy_content_to_temp();
+
+            // Extract the archive to the required dir.
+            $zip = new \ZipArchive();
+            $zip->open($temppath);
+            $zip->extractTo($target);
```

Comment:
possibly doesn't matter but the old code does the chmod 0777, curious if the extract ends up the same

Reply:
This doesnt seem to matter, The end result inside the zipped archive all ends up as 664 as far as I can see anyway